### ISSUE TITLE
[GAME] Tiles für pathfinding verwenden

### ISSUE THREAD

**Issue by AMatutat** [Programmiermethoden/Dungeon#787]
@AMatutat Ich würde sogar noch einen Schritt weitergehen und tatsächlich `Optional<Tile>` zurückliefern und dem Aufrufer überlassen, sich die Coordinate oder den Point da rauszuprökeln. Also `public static Optional<Tile> randomAccessibleTileInRange(` ... Liest sich auch gleich viel besser :)

_Originally posted by @cagix in https://github.com/Programmiermethoden/Dungeon/pull/775#discussion_r1239810008_

**Comment by AMatutat** [comment]
Da will ich nochmal drüber grübeln, solange wimi-label

**Comment by AMatutat** [comment]
Mal dumm gefragt, wenn wir `Point extends Coordinate` machen würden, könnten wir uns den ganzen mist nicht viel viel einfacher machen? 
Ich weiß grade nur nicht ob das so geht wie ich mir das vorstelel, aber theoretisch könnten wir dann überall wo wir eine Coordinate brauchen auch ein Point reinstecken und bräuchten keine API mehr dafür

**Comment by cagix** [comment]
Hmmm, fies. :). Das ist die alte Diskussion: Wollen wir punktgenau arbeiten können oder kachelbasiert? 

Nach meinem Verständnis ist die Benutzung von `Tile` und `Coordinate` relativ synonym, oder? Dann bräuchte man ja nur noch beim Kunden `Point` und/oder `Tile`. Wenn die beiden Klassen ein gemeinsames Interface bekommen könnten, könnte die `PositionComponent` tatsächlich beides speichern und es wäre für den Kunden erstmal egal bzw. davon abhängig, was er wirklich reintut.

**Comment by cagix** [comment]
> Mal dumm gefragt, wenn wir `Point extends Coordinate` machen würden, könnten wir uns den ganzen mist nicht viel viel einfacher machen?
> Ich weiß grade nur nicht ob das so geht wie ich mir das vorstelel, aber theoretisch könnten wir dann überall wo wir eine Coordinate brauchen auch ein Point reinstecken und bräuchten keine API mehr dafür

🤣Meine Rede :)

**Comment by AMatutat** [comment]
Ämm @cagix ich bin verwirrt.
Ich hab Polymorphe irgendwie anders in erinnerung:

```java
public class Main {
    public static void main(String[] args) throws IOException {
        X x = new X(1, 2);
        Y y = new Y(1.5f, 3.2f);
        X z = new Y(1.3f, 3.2f);
        System.out.println(x.a);
        System.out.println(x.b);
        System.out.println(y.a);
        System.out.println(y.b);
        System.out.println(z.a);
        System.out.println(z.b);
    }

    public static class X {
        public int a;
        public int b;

        public X(int a, int b) {
            this.a = a;
            this.b = b;
        }
    }

    public static class Y extends X {
        public float a;
        public float b;

        public Y(float a, float b) {
            super((int) a, (int) b);
            this.a = a;
            this.b = b;
        }
    }
}
```

Ausgabe ist 
```
1
2
1.5
3.2
1
3
```

**Comment by AMatutat** [comment]
> > Mal dumm gefragt, wenn wir `Point extends Coordinate` machen würden, könnten wir uns den ganzen mist nicht viel viel einfacher machen?
> > Ich weiß grade nur nicht ob das so geht wie ich mir das vorstelel, aber theoretisch könnten wir dann überall wo wir eine Coordinate brauchen auch ein Point reinstecken und bräuchten keine API mehr dafür
> 
> 🤣Meine Rede :)
> 
> Edit: `Coordinate` wird das neue Interface, und `Point implements Coordinate` und aber auch `Tile implements Coordindate`. Dann brauchst Du auf Kundenseite nur noch `Coordinate`.
> 
> Edit: Und natürlich `Coordinate.of(Point)` und `Coordinate.of(Tile)`, damit der Kunde auf Wunsch punktgenau arbeiten kann.

hmm, an ein interface habe ich nie gedacht. 
Das muss ich nochmal kritisch untersuchen

**Comment by cagix** [comment]
> Ämm @cagix ich bin verwirrt. Ich hab Polymorphe irgendwie anders in erinnerung:
> 
> ```java
> public class Main {
>     public static void main(String[] args) throws IOException {
>         X x = new X(1, 2);
>         Y y = new Y(1.5f, 3.2f);
>         X z = new Y(1.3f, 3.2f);
>         System.out.println(x.a);
>         System.out.println(x.b);
>         System.out.println(y.a);
>         System.out.println(y.b);
>         System.out.println(z.a);
>         System.out.println(z.b);
>     }
> 
>     public static class X {
>         public int a;
>         public int b;
> 
>         public X(int a, int b) {
>             this.a = a;
>             this.b = b;
>         }
>     }
> 
>     public static class Y extends X {
>         public float a;
>         public float b;
> 
>         public Y(float a, float b) {
>             super((int) a, (int) b);
>             this.a = a;
>             this.b = b;
>         }
>     }
> }
> ```
> 
> Ausgabe ist
> 
> ```
> 1
> 2
> 1.5
> 3.2
> 1
> 3
> ```

was genau verwirrt dich grad? z? mach mal den zugriff über eine methode, dann sollte es anders aussehen.

**Comment by cagix** [comment]
> > > Mal dumm gefragt, wenn wir `Point extends Coordinate` machen würden, könnten wir uns den ganzen mist nicht viel viel einfacher machen?
> > > Ich weiß grade nur nicht ob das so geht wie ich mir das vorstelel, aber theoretisch könnten wir dann überall wo wir eine Coordinate brauchen auch ein Point reinstecken und bräuchten keine API mehr dafür
> > 
> > 
> > 🤣Meine Rede :)
> > Edit: `Coordinate` wird das neue Interface, und `Point implements Coordinate` und aber auch `Tile implements Coordindate`. Dann brauchst Du auf Kundenseite nur noch `Coordinate`.
> > Edit: Und natürlich `Coordinate.of(Point)` und `Coordinate.of(Tile)`, damit der Kunde auf Wunsch punktgenau arbeiten kann.
> 
> hmm, an ein interface habe ich nie gedacht. Das muss ich nochmal kritisch untersuchen

Und damit es kürzer wird ("Coordinate" ist echt viel zu tippen) und nicht zu Verwirrungen mit Alt-Eingesessenen führt, ist vermutlich `Position` ein besserer Name für das gemeinsame Interface :)

**Comment by AMatutat** [comment]
> > Ämm @cagix ich bin verwirrt. Ich hab Polymorphe irgendwie anders in erinnerung:
> > ```java
> > public class Main {
> >     public static void main(String[] args) throws IOException {
> >         X x = new X(1, 2);
> >         Y y = new Y(1.5f, 3.2f);
> >         X z = new Y(1.3f, 3.2f);
> >         System.out.println(x.a);
> >         System.out.println(x.b);
> >         System.out.println(y.a);
> >         System.out.println(y.b);
> >         System.out.println(z.a);
> >         System.out.println(z.b);
> >     }
> > 
> >     public static class X {
> >         public int a;
> >         public int b;
> > 
> >         public X(int a, int b) {
> >             this.a = a;
> >             this.b = b;
> >         }
> >     }
> > 
> >     public static class Y extends X {
> >         public float a;
> >         public float b;
> > 
> >         public Y(float a, float b) {
> >             super((int) a, (int) b);
> >             this.a = a;
> >             this.b = b;
> >         }
> >     }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Ausgabe ist
> > ```
> > 1
> > 2
> > 1.5
> > 3.2
> > 1
> > 3
> > ```
> 
> was genau verwirrt dich grad? z? mach mal den zugriff über eine methode, dann sollte es anders aussehen.

Genau, `z`. Klar über einen richtigen `getter` würde ich den `Y` kram aufrufen. Aber ich versteh nicht ganz, warum das für Attribute nicht gilt. 
Ich hätte jetzt analoges verhalten erwartet und daher für. `z` die Ausgaben 1.3 und 3.2
Hardcore Off-Topic. Vielleicht muss ich mir das nach einer Tasse Kaffe nochmal anschauen.

**Comment by cagix** [comment]
> Genau, z. Klar über einen richtigen getter würde ich den Y kram aufrufen. Aber ich versteh nicht ganz, warum das für Attribute nicht gilt.

Dynamische Polymorphie ist in Java Methoden-zentriert. Ich schaue zur Laufzeit beim Aufruf einer Methode nach, welchen Typ ich wirklich habe und gehe dann schrittweise "nach oben",  um die am tiefsten in der Hierarchie implementierte Methode zu finden. Bzgl. der Attribute sagt die Laufzeitumgebung: Hmmm, das ist vom Typ her ein X, also nehme die Attribute aus diesem Teil - und in Deinem Modell überdecken sich die Attribute von den Namen her.

**Comment by cagix** [comment]
geile frage für die klausur :) 👍

**Comment by cagix** [comment]
Dieses Ticket erscheint als eine Detail-Ausprägung von #547 (hier mit dem Aspekt "Pathfinding").

**Comment by AMatutat** [comment]
Ich mach hier mal als duplicate zu.
Die Diskussionen sollten in #547 aber mit bedacht werden.
